<img width="536" height="164" alt="image" src="https://github.com/user-attachments/assets/3ae15e48-9971-4176-993b-d062264d40b4" />

콘솔 로그 : 
D:\compasstep\Compasstep\Compasstep-BackEnd\src\main\java\com\fifo\compasstep\admin\service\AdminService.java:13: error: package com.fifo.compasstep.chat.repository does not exist
import com.fifo.compasstep.chat.repository.ChatRepository;

해당 오류가 발생합니다. 
chat 도메인의 repository가 없는 것 같습니다.

--- a/src/main/java/com/fifo/compasstep/admin/controller/AdminController.java
+++ b/src/main/java/com/fifo/compasstep/admin/controller/AdminController.java
@@ -14,6 +14,8 @@ import org.springframework.security.core.annotation.AuthenticationPrincipal;
 import org.springframework.validation.annotation.Validated;
 import org.springframework.web.bind.annotation.*;
 
+import java.util.List;
+
 @RestController
 @RequestMapping("/api/admin")
 @RequiredArgsConstructor
@@ -88,4 +90,27 @@ public class AdminController {
         return ApiResponse.success(result);
     }
 
+    @GetMapping("/users/malicious")
+    public ApiResponse<List<AdminResponseDTO.MaliciousResponseDTO>> getMaliciousUsers() {
+        List<AdminResponseDTO.MaliciousResponseDTO> result = adminService.getMaliciousUsers();
+        return  ApiResponse.success(result);
+    }
+
+    @GetMapping("users/logs/{userPKId}")
+    public ApiResponse<AdminResponseDTO.ChatLogsListDTO> getUserChatLogs(@PathVariable Long userPKId) {
+        AdminResponseDTO.ChatLogsListDTO result = adminService.getUserChatLogs(userPKId);
+        return ApiResponse.success(result);
+    }
+
+    @PatchMapping("/users/ban/{userPKId}")
+    public ApiResponse<Void> banUser(@PathVariable Long userPKId) {
+        adminService.banUser(userPKId);
+        return ApiResponse.success(null);
+    }
+
+    @PatchMapping("/users/unban/{userPKId}")
+    public ApiResponse<Void> unbanUser(@PathVariable Long userPKId) {
+        adminService.unbanUser(userPKId);
+        return ApiResponse.success(null);
+    }
 }

--- a/src/main/java/com/fifo/compasstep/admin/dto/AdminResponseDTO.java
+++ b/src/main/java/com/fifo/compasstep/admin/dto/AdminResponseDTO.java
@@ -1,10 +1,13 @@
 package com.fifo.compasstep.admin.dto;
 
+import com.fifo.compasstep.user.enums.Status;
 import lombok.AllArgsConstructor;
 import lombok.Builder;
 import lombok.Getter;
 import lombok.NoArgsConstructor;
+import org.checkerframework.checker.units.qual.A;
 
+import java.time.LocalDateTime;
 import java.util.List;
 
 public class AdminResponseDTO {
@@ -30,6 +33,37 @@ public class AdminResponseDTO {
     public static class AdminInfoDTO {
         private String nickname;
         private String email;
+    }
+
+    @Getter
+    @Builder
+    @NoArgsConstructor
+    @AllArgsConstructor
+    public static class MaliciousResponseDTO {
+        private Long userId;
+        private String name;
+        private String email;
+        private Status status;
+        private LocalDateTime createdAt;
+    }
+
+    @Getter
+    @Builder
+    @NoArgsConstructor
+    @AllArgsConstructor
+    public static class ChatLogsListDTO{
+        private List<ChatLogs> chatLogs;
+    }
+
+    @Getter
+    @Builder
+    @NoArgsConstructor
+    @AllArgsConstructor
+    public static class ChatLogs {
+        private Long chatId;
+        private LocalDateTime createdAt;
+        private String content;
+        private Boolean isGuardrailed;
     }
 
 }

--- a/src/main/java/com/fifo/compasstep/admin/exceptions/AdminErrorStatus.java
+++ b/src/main/java/com/fifo/compasstep/admin/exceptions/AdminErrorStatus.java
@@ -17,6 +17,7 @@ public enum AdminErrorStatus implements BaseErrorCode {
     ANOTHER_ERROR(HttpStatus.NOT_FOUND, 500, "다른 오류."),
     ADMIN_NOT_FOUND(HttpStatus.NOT_FOUND, 403, "관리자가 없습니다."),
     ADMIN_LOGIN_FAILED(HttpStatus.UNAUTHORIZED, 401, "아이디 혹은 비밀번호가 틀렸습니다."),
+    ACCESS_DENIED(HttpStatus.UNAUTHORIZED, 403, "권한이 없어 요청이 거부되었습니다"),
     REDIS_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, 500, "Redis 오류가 발생했습니다."),
 
     REFRESH_TOKEN_NOT_FOUND(HttpStatus.NOT_FOUND, 401, "리프레시 토큰을 찾을 수 없습니다."),

--- a/src/main/java/com/fifo/compasstep/admin/service/AdminService.java
+++ b/src/main/java/com/fifo/compasstep/admin/service/AdminService.java
@@ -8,6 +8,9 @@ import com.fifo.compasstep.admin.exceptions.AdminErrorStatus;
 import com.fifo.compasstep.admin.repository.AdminRepository;
 import com.fifo.compasstep.apipayload.ApiResponse;
 import com.fifo.compasstep.apipayload.exceptions.handler.AdminHandler;
+import com.fifo.compasstep.apipayload.exceptions.handler.UserHandler;
+import com.fifo.compasstep.chat.domain.Chat;
+import com.fifo.compasstep.chat.repository.ChatRepository;
 import com.fifo.compasstep.security.jwt.JwtProperties;
 import com.fifo.compasstep.security.jwt.JwtTokenProvider;
 import com.fifo.compasstep.security.service.RefreshTokenService;
@@ -15,6 +18,8 @@ import com.fifo.compasstep.security.userDetails.AdminUserDetails;
 import com.fifo.compasstep.security.userDetails.UserUserDetails; // UserUserDetails import
 import com.fifo.compasstep.security.util.CookieUtil;
 import com.fifo.compasstep.user.domain.User; // User import
+import com.fifo.compasstep.user.enums.Status;
+import com.fifo.compasstep.user.exceptions.UserErrorStatus;
 import com.fifo.compasstep.user.repository.UserRepository; // UserRepository import
 import jakarta.servlet.http.HttpServletRequest;
 import jakarta.servlet.http.HttpServletResponse;
@@ -47,6 +52,7 @@ public class AdminService {
     private final PasswordEncoder passwordEncoder;
     private final EmailService emailService;
     private final UserRepository userRepository; // refreshToken 로직에서 User 조회용
+    private final ChatRepository chatRepository;
 
     @Transactional
     public AdminResponseDTO.LoginResponseDTO login(AdminRequestDTO.AdminLoginRequestDTO request, HttpServletResponse response) {
@@ -292,6 +298,63 @@ public class AdminService {
         return AdminResponseDTO.AdminListResponseDTO.builder()
                 .adminList(adminInfoDTOs)
                 .build();
+    }
+
+    @Transactional
+    public List<AdminResponseDTO.MaliciousResponseDTO> getMaliciousUsers() {
+        List<User> maliciousUser = userRepository.findByStatusIn(
+                List.of(Status.SUSPENDED, Status.DELETED)
+        );
+
+        return maliciousUser.stream()
+                .map(user -> AdminResponseDTO.MaliciousResponseDTO.builder()
+                        .userId(user.getId())
+                        .name(user.getName())
+                        .email(user.getEmail())
+                        .status(user.getStatus())
+                        .createdAt(user.getCreatedAt()) // BaseEntity의 createdAt 사용
+                        .build())
+                .collect(Collectors.toList());
+    }
+
+    @Transactional
+    public AdminResponseDTO.ChatLogsListDTO getUserChatLogs(Long userPKId) {
+        // 1. userId로 User 엔티티 조회 (없으면 예외 발생)
+        User user = userRepository.findById(userPKId)
+                .orElseThrow(() -> new UserHandler(UserErrorStatus.USER_NOT_FOUND));
+
+        // 2. ChatRepository를 이용해 해당 User의 최신 채팅 20개 조회
+        List<Chat> recentChats = chatRepository.findTop20ByUserOrderByCreatedAtDesc(user);
+
+        // 3. Chat 엔티티 List를 ChatLogs DTO List로 변환
+        List<AdminResponseDTO.ChatLogs> chatLogsList = recentChats.stream()
+                .map(chat -> AdminResponseDTO.ChatLogs.builder()
+                        .chatId(chat.getId())
+                        .createdAt(chat.getCreatedAt())
+                        .content(chat.getContent())
+                        .isGuardrailed(chat.getIsGuardrailed())
+                        .build())
+                .collect(Collectors.toList());
+        // 4. 최종 ChatLogsListDTO로 감싸서 반환
+        return AdminResponseDTO.ChatLogsListDTO.builder()
+                .chatLogs(chatLogsList)
+                .build();
+
+    }
+
+    @Transactional
+    public void banUser(Long userPKId) {
+        User user = userRepository.findById(userPKId)
+                .orElseThrow(() -> new UserHandler(UserErrorStatus.USER_NOT_FOUND));
+
+        user.changeStatus(Status.BLOCKED);
+    }
+
+    @Transactional
+    public void unbanUser(Long userPKId) {
+        User user = userRepository.findById(userPKId)
+                .orElseThrow(() -> new UserHandler(UserErrorStatus.USER_NOT_FOUND));
+        user.changeStatus(Status.NORMAL);
     }
 
     private Map<String, String> makeTemPassword() {

--- a/src/main/java/com/fifo/compasstep/admin/service/AdminService.java
+++ b/src/main/java/com/fifo/compasstep/admin/service/AdminService.java
@@ -303,7 +303,7 @@ public class AdminService {
     @Transactional
     public List<AdminResponseDTO.MaliciousResponseDTO> getMaliciousUsers() {
         List<User> maliciousUser = userRepository.findByStatusIn(
-                List.of(Status.SUSPENDED, Status.DELETED)
+                List.of(Status.SUSPENDED, Status.BLOCKED)
         );
 
         return maliciousUser.stream()

--- a/src/main/java/com/fifo/compasstep/apipayload/exceptions/ExceptionAdvice.java
+++ b/src/main/java/com/fifo/compasstep/apipayload/exceptions/ExceptionAdvice.java
@@ -1,6 +1,8 @@
 package com.fifo.compasstep.apipayload.exceptions;
 
+import com.fifo.compasstep.admin.exceptions.AdminErrorStatus;
 import com.fifo.compasstep.apipayload.ApiResponse;
+import com.fifo.compasstep.apipayload.code.BaseErrorCode;
 import com.fifo.compasstep.apipayload.code.ErrorReasonDTO;
 import com.fifo.compasstep.reference.exceptions.ReferenceErrorStatus;
 import jakarta.servlet.http.HttpServletRequest;
@@ -14,6 +16,7 @@ import org.springframework.web.context.request.ServletWebRequest;
 import org.springframework.web.context.request.WebRequest;
 import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
 
+import java.nio.file.AccessDeniedException;
 import java.security.GeneralSecurityException;
 
 @Slf4j
@@ -51,12 +54,29 @@ public class ExceptionAdvice extends ResponseEntityExceptionHandler {
         return handleExceptionInternal(e, r, null, request);
     }
 
-    /** 마지막 안전망 → 500 */
-    @ExceptionHandler(Exception.class)
-    @SuppressWarnings("unused")
-    public ResponseEntity<Object> handleUnknown(Exception e, HttpServletRequest request) {
-        log.error("Unhandled exception", e);
-        var r = ReferenceErrorStatus.EXTERNAL_API_ERROR.getReasonHttpStatus();
-        return handleExceptionInternal(e, r, null, request);
+    @ExceptionHandler(AccessDeniedException.class)
+    public ResponseEntity<Object> handleAccessDenied(AccessDeniedException ex, HttpServletRequest request) {
+        // 1. 미리 정의된 에러 코드/상태를 가져옵니다 (예: SecurityErrorStatus.ACCESS_DENIED)
+        BaseErrorCode errorCode = AdminErrorStatus.ACCESS_DENIED; // ACCESS_DENIED를 SecurityErrorStatus에 정의해야 함
+
+        // 2. ApiResponse 형식으로 실패 응답 본문을 만듭니다.
+        ApiResponse<Object> body = ApiResponse.onFailure(
+                errorCode.getReason().getCode(),    // 예: "AUTH003"
+                errorCode.getReason().getMessage(), // 예: "접근 권한이 없습니다."
+                null                               // 추가 데이터 없음
+        );
+
+        // 3. ResponseEntity에 본문과 HTTP 상태 코드(403 Forbidden)를 담아 반환합니다.
+        return ResponseEntity.status(errorCode.getReasonHttpStatus().getHttpStatus()) // HttpStatus.FORBIDDEN
+                .body(body);
     }
+
+//    /** 마지막 안전망 → 500 */
+//    @ExceptionHandler(Exception.class)
+//    @SuppressWarnings("unused")
+//    public ResponseEntity<Object> handleUnknown(Exception e, HttpServletRequest request) {
+//        log.error("Unhandled exception", e);
+//        var r = ReferenceErrorStatus.EXTERNAL_API_ERROR.getReasonHttpStatus();
+//        return handleExceptionInternal(e, r, null, request);
+//    }
 }

--- a/src/main/java/com/fifo/compasstep/reference/controller/DiscoveryController.java
+++ b/src/main/java/com/fifo/compasstep/reference/controller/DiscoveryController.java
@@ -8,6 +8,7 @@ import com.fifo.compasstep.reference.service.DiscoveryService;
 import com.fifo.compasstep.security.userDetails.UserUserDetails;
 import jakarta.validation.Valid;
 import lombok.RequiredArgsConstructor;
+import org.springframework.security.access.prepost.PreAuthorize;
 import org.springframework.security.core.annotation.AuthenticationPrincipal;
 import org.springframework.web.bind.annotation.*;
 
@@ -22,6 +23,7 @@ public class DiscoveryController {
 
     /** 키워드 기반 레퍼런스 탐색 (FastAPI 프록시) */
     @PostMapping("/discovery/keyword")
+    @PreAuthorize("hasAuthority('STATUS_NORMAL')")
     public ApiResponse<List<TrackVideoDto>> discoveryKeyword(
             @Valid @RequestBody DiscoveryKeywordRequestDto req,
             @AuthenticationPrincipal UserUserDetails currentUser

--- a/src/main/java/com/fifo/compasstep/security/config/CustomAccessDenyHandler.java
+++ b/src/main/java/com/fifo/compasstep/security/config/CustomAccessDenyHandler.java
@@ -1,0 +1,41 @@
+package com.fifo.compasstep.security.config;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fifo.compasstep.admin.exceptions.AdminErrorStatus;
+import com.fifo.compasstep.apipayload.ApiResponse;
+import com.fifo.compasstep.apipayload.code.BaseErrorCode;
+import com.fifo.compasstep.security.exceptions.SecurityErrorStatus;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.web.access.AccessDeniedHandler;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+
+@Component
+@RequiredArgsConstructor
+public class CustomAccessDenyHandler implements AccessDeniedHandler {
+
+    private final ObjectMapper objectMapper;
+
+    @Override
+    public void handle(HttpServletRequest request,
+                       HttpServletResponse response,
+                       AccessDeniedException accessDeniedException) throws IOException, ServletException {
+
+        BaseErrorCode errorCode = AdminErrorStatus.ACCESS_DENIED; // 403 에러 코드 사용
+
+        ApiResponse<Object> apiResponse = ApiResponse.onFailure(
+                errorCode.getReason().getCode(),
+                errorCode.getReason().getMessage(),
+                null
+        );
+
+        response.setStatus(errorCode.getReasonHttpStatus().getHttpStatus().value()); // 403 상태 코드
+        response.setContentType("application/json;charset=UTF-8");
+        response.getWriter().write(objectMapper.writeValueAsString(apiResponse));
+    }
+}

--- a/src/main/java/com/fifo/compasstep/security/config/SecurityConfig.java
+++ b/src/main/java/com/fifo/compasstep/security/config/SecurityConfig.java
@@ -28,6 +28,7 @@ public class SecurityConfig {
     private final JwtAuthenticationFilter jwtAuthenticationFilter;
     private final AdminUserDetailsService adminUserDetailsService;
     private final JwtAuthenticationEntryPoint jwtAuthenticationEntryPoint;
+    private final CustomAccessDenyHandler customAccessDeniedHandler;
 
     @Bean
     public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
@@ -36,9 +37,11 @@ public class SecurityConfig {
                 .sessionManagement()
                 .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
                 .and()
-                .exceptionHandling()
-                .authenticationEntryPoint(jwtAuthenticationEntryPoint)
-                .and()
+                .exceptionHandling(ex -> ex
+                        .authenticationEntryPoint(jwtAuthenticationEntryPoint) // 인증 안 된 접근 -> 401
+                        .accessDeniedHandler(customAccessDeniedHandler)     // 인증 됐지만 권한 부족 -> 403 <-- 추가!
+                )
+                
                 .authorizeHttpRequests(auth -> auth
                         //로그인 밑 회원가입 경로 전부 허용
                         .requestMatchers("/api/admin/login", "/api/admin/logout", "/auth/oauth/**", "/api/admin/refresh",

--- a/src/main/java/com/fifo/compasstep/security/config/SecurityConfig.java
+++ b/src/main/java/com/fifo/compasstep/security/config/SecurityConfig.java
@@ -10,6 +10,7 @@ import org.springframework.security.authentication.AuthenticationManager;
 import org.springframework.security.authentication.AuthenticationProvider;
 import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
 import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
 import org.springframework.security.config.annotation.web.builders.HttpSecurity;
 import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
 import org.springframework.security.config.http.SessionCreationPolicy;
@@ -21,6 +22,7 @@ import org.springframework.security.web.authentication.UsernamePasswordAuthentic
 @Configuration
 @EnableWebSecurity
 @RequiredArgsConstructor
+@EnableMethodSecurity(prePostEnabled = true)
 public class SecurityConfig {
 
     private final JwtAuthenticationFilter jwtAuthenticationFilter;

